### PR TITLE
[SimplifyCFG] Consider preds to switch in `simplifyDuplicateSwitchArms`

### PR DESCRIPTION
**Pull request by antoniofrighetto**
Allow a duplicate basic block with multiple predecessors to the jump table to be simplified, by considering that the same basic block may appear in more switch cases.

### CODE REVIEW

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7577]: Use `hasNPredecessorsOrMore(4)` and `BB->getUniquePredecessor() == SI->getParent()`.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (michaelmaitland) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7577]: What is special about having >= 4 predecessors?

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll
Comment (michaelmaitland) [inline llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll:343]: Did you mean to change these lines?

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7577]: It is used to avoid `O(N^2)` lookup. We may choose a larger threshold if the compile-time impact is negligible.

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll
Comment (antoniofrighetto) [reply llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll:343]: I took the chance to regenerate the test too, as the `if.else` branch was dead.

**Review by nikic — COMMENTED** [review]
Can you please pre-commit the tests and add a test where the transform triggers but does not convert into a select afterwards (only reduces blocks)?

**Review by michaelmaitland — APPROVED** [review]
LGTM

**Pull request merged**
